### ISSUE TITLE
Writeup?

### ISSUE THREAD

**Issue by Naville** [JuliaPoo/Artfuscator#2]
Hey:
I've been patiently waiting for the "How it works" sections to be filled but seems not?
Is there any write-up or material on this topic as I'm looking to reproduce similar stuff on LLVM.

Thanks!

**Comment by Naville** [comment]
https://www.youtube.com/watch?v=HlUe0TUHOIc and https://github.com/Battelle/REpsych/blob/master/slides/domas_defcon_2015_repsych.pdf , it seems

**Comment by JuliaPoo** [comment]
I wrote this repository pretty much over the course of 1-2 days so
1. It's not much
2. Code quality is pretty terrible

I totally didn't expect the amount of attention it got, and I've been lacking the motivation to write a writeup.

With that out of the way, how Artfuscator works is pretty simple actually. At a high level, Artfuscator lies to the disassembler about its control flow, adding a bunch of meaningless jumps and switch cases to form the image. The actual control flow is handled via `call` instructions. The main ELVM->x86 code can be found [here](https://github.com/JuliaPoo/elvm/blob/art-backend/target/art.c), which emits a nasm file that basically looks like one giant basic block to the disassembler (which doesn't consider `call` instructions as BB exits). There's a [python script](https://github.com/JuliaPoo/Artfuscator/blob/main/artfuscator/utils.py) that then chops up the giant BB to create the image. Also, reversing the actual binary to figure out how it is compiled would probably be helpful too.

If you do intend to implement this from LLVM, it would be very helpful to only use a subset of the registers available, so that you have a bunch of registers for the "boilerplate" code to trick the disassembler. In Artfuscator, an exception is the `EFLAG` register which is used by everything pretty much, and I had to be careful when inserting jumps between my actual instructions so that the extra jumps don't affect the correct controlflow. This kind of info can be retrieved from a disassembler like Intel XED.
